include/pybind11/pybind11.h

```diff
@@ -504,15 +504,15 @@ class cpp_function : public function {
 
         auto self_value_and_holder = value_and_holder();
         if (overloads->is_constructor) {
-            const auto tinfo = get_type_info((PyTypeObject *) overloads->scope.ptr());
-            const auto pi = reinterpret_cast<instance *>(parent.ptr());
```

Comment:
Comments from #2746:

- @bstaletic: Without checking the type first, we were reinterpreting `NotPybindDerived()` as `instance*`.
- @YannickJadoul: See also the code below that gets simplified; `cfr. if (!self_value_and_holder.type || !self_value_and_holder.inst) -> pi->get_value_and_holder(tinfo, true);`